### ISSUE TITLE
samtools throws stderr / .bam file is not populated

### ISSUE THREAD

**Issue by stevechen123** [CenikLab/satmut_utils#2]
Hi. Thanks for creating the satmut_utils. 

Perhaps similar to the previous issue posted ([https://github.com/CenikLab/satmut_utils/issues/1], I am stuck at the step where Samtools throwings an issue saying that the .bam is in a format that cannot be usefully indexed error. The issue might be associated with Bowtie2 (the immediate preceding step) or Samtools itself. The information below might be useful for diagnosing the problem.

**Program versions**
satmut_utils version 1.0.3
Bowtie 2 version 2.5.1
samtools version 1.17

**Running the unit test results:**
[unit_test.txt](https://github.com/CenikLab/satmut_utils/files/11884922/unit_test.txt)

Ran 175 tests in 17.131s

FAILED (errors=10)

**Reference files**
[Reference_files.zip](https://github.com/CenikLab/satmut_utils/files/11884722/Reference_files.zip)

**Command ran:**
satmut_utils -r ./references/HuRhWT_yms.fa call -1 F300010069_L01_25_Synlin20CycleKODAmp+Pur_BC128_1.fq.gz -2 F300010069_L01_25_Synlin20CycleKODAmp+Pur_BC128_1_rc.fq.gz -g ./references/HuRhWT_yms.gff -k ./references/HuRhWT_yms.fa --keep_intermediates

**Terminal output**
2023-06-27 11:52:02,056 - root - INFO - Started /home/steven/anaconda3/envs/satmut_utils/bin/satmut_utils
2023-06-27 11:52:02,056 - root - INFO - Starting call workflow.
2023-06-27 11:52:02,056 - root - INFO - Copying input reference FASTA and indexing.
2023-06-27 11:52:02,058 - analysis.references - INFO - Generating FASTA index file for ./HuRhWT_yms.fa.
2023-06-27 11:52:02,101 - analysis.references - INFO - Building FM index files for ./HuRhWT_yms.fa.
2023-06-27 11:52:02,800 - root - INFO - Indexing the GFF reference FASTA.
2023-06-27 11:52:02,836 - analysis.read_preprocessor - INFO - Started FASTQ preprocessor workflow.
2023-06-27 11:52:02,837 - analysis.read_preprocessor - INFO - Started cutadapt trimming.
2023-06-27 11:53:49,089 - analysis.read_preprocessor - INFO - Completed cutadapt trimming.
2023-06-27 11:53:49,089 - analysis.read_preprocessor - INFO - Completed FASTQ preprocessor workflow.
2023-06-27 11:53:49,090 - analysis.aligners - INFO - Started bowtie2 aligner workflow.
2023-06-27 11:53:49,090 - analysis.aligners - INFO - Writing output BAM /media/scratch/stevenscratch/Syn_lib_seq_data/2023_06_26_synlib_raw_lib/F300010069_L01_25_Synlin20CycleKODAmp+Pur_BC128_1.bam
Traceback (most recent call last):
  File "/home/steven/anaconda3/envs/satmut_utils/bin/satmut_utils", line 8, in <module>
    sys.exit(main())
  File "/home/steven/anaconda3/envs/satmut_utils/lib/python3.7/site-packages/satmut_utils/satmut_utils.py", line 526, in main
    keep_intermediates=args_dict["keep_intermediates"])
  File "/home/steven/anaconda3/envs/satmut_utils/lib/python3.7/site-packages/satmut_utils/satmut_utils.py", line 435, in call_workflow
    nthreads=bowtie2_nthreads)
  File "/home/steven/anaconda3/envs/satmut_utils/lib/python3.7/site-packages/scripts/run_bowtie2_aligner.py", line 85, in workflow
    bt = Bowtie2(config=bc, f1=f1_full, f2=f2_full, output_dir=outdir_full, output_bam=outbam)
  File "/home/steven/anaconda3/envs/satmut_utils/lib/python3.7/site-packages/analysis/aligners.py", line 117, in __init__
    self.workflow()
  File "/home/steven/anaconda3/envs/satmut_utils/lib/python3.7/site-packages/analysis/aligners.py", line 184, in workflow
    su.index_bam(self.output_bam)
  File "/home/steven/anaconda3/envs/satmut_utils/lib/python3.7/site-packages/analysis/seq_utils.py", line 328, in index_bam
    pysam.index(bam)
  File "/home/steven/anaconda3/envs/satmut_utils/lib/python3.7/site-packages/pysam/utils.py", line 89, in __call__
    stderr))
pysam.utils.SamtoolsError: 'samtools returned with error 1: stdout=, stderr=samtools index: "/media/scratch/stevenscratch/Syn_lib_seq_data/2023_06_26_synlib_raw_lib/F300010069_L01_25_Synlin20CycleKODAmp+Pur_BC128_1.bam" is in a format that cannot be usefully indexed\n'

**Output files**
<img width="681" alt="Screenshot 2023-06-27 at 11 59 58 AM" src="https://github.com/CenikLab/satmut_utils/assets/75347745/9a3afb20-82bf-4ee0-964f-1b94c4d6ea11">

[satmut_utils.log](https://github.com/CenikLab/satmut_utils/files/11884799/satmut_utils.log)
[Bowtie2.stderr.log](https://github.com/CenikLab/satmut_utils/files/11884800/Bowtie2.stderr.log)

**/tmp files generated if run without --keep_intermediates**
<img width="634" alt="Screenshot 2023-06-27 at 12 04 00 PM" src="https://github.com/CenikLab/satmut_utils/assets/75347745/dd38f66d-3b6b-4baa-a7f0-5bdb3bb473e1">

The .bam file is empty but the cutadapt .trimmed outputs are populated.

**Hardware**
This was run on a desktop computer. AMD 5950x 16 cores/32 threads with 128 GB DDR4 RAM.

**Additional comments**
I also tried to download on my Mac using the installation instructions but it would not generate the conda environment.

Any help is appreciated!

**Comment by ijhoskins** [comment]
Hi @stevechen123 thanks for creating an issue. Your reference files look fine so that is unlikely the issue.

The Bowtie2 log indicates the issue is a failure to resolve the reference FASTA path, leading to a malformed bowtie2 call. It is not clear to me why this is the case as the reference FASTA is copied to the output directory and used for subsequent alignment. If you pass a full expanded path does that solve the issue?

I also noticed your trimmed FASTQs are of larger size than the input FASTQs, which is unexpected. Can you provide some test FASTQs so I can attempt to reproduce the error? 40-400 lines from each input FASTQ should be plenty.

**Comment by stevechen123** [comment]
Hi @ijhoskins. I tried passing the full expanded path yesterday and again just now using the following command but it continues to throw the same error.

satmut_utils -r /media/scratch/stevenscratch/Syn_lib_seq_data/2023_06_26_synlib_raw_lib/references/HuRhWT_yms.fa call -1 F300010069_L01_25_Synlin20CycleKODAmp+Pur_BC128_1.fq.gz -2 F300010069_L01_25_Synlin20CycleKODAmp+Pur_BC128_1_rc.fq.gz -g /media/scratch/stevenscratch/Syn_lib_seq_data/2023_06_26_synlib_raw_lib/references/HuRhWT_yms.gff -k /media/scratch/stevenscratch/Syn_lib_seq_data/2023_06_26_synlib_raw_lib/references/HuRhWT_yms.fa

I've attached 400 lines of my FASTQ files here for you to test. The file labelled with "rc" is just a reverse complement of the first—given my usecase. Thanks! 🙏 

[fastq_files_400lines.zip](https://github.com/CenikLab/satmut_utils/files/11886172/fastq_files_400lines.zip)

**Comment by ijhoskins** [comment]
Hi @stevechen123 

I was able to successfully run your test FASTQs on two different Linux servers:
```
satmut_utils -o $WORK2/HuRhWT/results -r $WORK2/HuRhWT/HuRhWT_yms.fa call -1 $WORK2/HuRhWT/F300010069_L01_25_Synlin20CycleKODAmp+Pur_BC128_1_400.fq.gz -2 $WORK2/HuRhWT/F300010069_L01_25_Synlin20CycleKODAmp+Pur_BC128_1_rc_400.fq.gz -g $WORK2/HuRhWT/HuRhWT_yms.gff -k $WORK2/HuRhWT/HuRhWT_yms.fa --keep_intermediates
2023-06-27 14:57:59,027 - root - INFO - Started /home1/06285/ihoskins/miniconda3/envs/satmut_utils_test/bin/satmut_utils
2023-06-27 14:57:59,028 - root - INFO - Starting call workflow.
2023-06-27 14:57:59,029 - root - INFO - Copying input reference FASTA and indexing.
2023-06-27 14:57:59,032 - analysis.references - INFO - Generating FASTA index file for /work/06285/ihoskins/ls6/HuRhWT/results/HuRhWT_yms.fa.
2023-06-27 14:57:59,057 - analysis.references - INFO - Building FM index files for /work/06285/ihoskins/ls6/HuRhWT/results/HuRhWT_yms.fa.
2023-06-27 14:57:59,187 - analysis.read_preprocessor - INFO - Started FASTQ preprocessor workflow.
2023-06-27 14:57:59,187 - analysis.read_preprocessor - INFO - Started cutadapt trimming.
2023-06-27 14:57:59,418 - analysis.read_preprocessor - INFO - Completed cutadapt trimming.
2023-06-27 14:57:59,418 - analysis.read_preprocessor - INFO - Completed FASTQ preprocessor workflow.
2023-06-27 14:57:59,419 - analysis.aligners - INFO - Started bowtie2 aligner workflow.
2023-06-27 14:57:59,419 - analysis.aligners - INFO - Writing output BAM /work/06285/ihoskins/ls6/HuRhWT/results/F300010069_L01_25_Synlin20CycleKODAmp+Pur_BC128_1_.bam
2023-06-27 14:57:59,590 - analysis.aligners - INFO - Completed bowtie2 aligner workflow.
2023-06-27 14:57:59,590 - analysis.variant_caller - INFO - Initializing VariantCaller
2023-06-27 14:57:59,590 - analysis.read_preprocessor - INFO - Started variant call preprocessing workflow.
2023-06-27 14:57:59,590 - analysis.read_preprocessor - INFO - Sorting and splitting input BAM into R1 and R2.
2023-06-27 14:57:59,718 - analysis.read_preprocessor - INFO - Completed variant call preprocessing workflow.
2023-06-27 14:57:59,720 - analysis.variant_caller - INFO - Loading transcript CDS annotations for AA change determination.
2023-06-27 14:57:59,722 - analysis.coordinate_mapper - INFO - Collecting transcript CDS annotations.
2023-06-27 14:57:59,726 - analysis.coordinate_mapper - INFO - Pickling transcript CDS annotations to /work/06285/ihoskins/ls6/HuRhWT/results/HuRhWT_yms.cds.pkl.gz for /work/06285/ihoskins/ls6/HuRhWT/HuRhWT_yms.gff
2023-06-27 14:57:59,728 - analysis.variant_caller - INFO - Starting variant calling workflow.
[E::idx_find_and_load] Could not retrieve index file for '/work/06285/ihoskins/ls6/HuRhWT/results/F300010069_L01_25_Synlin20CycleKODAmp+Pur_BC128_1_.R1.call.bam'
[E::idx_find_and_load] Could not retrieve index file for '/work/06285/ihoskins/ls6/HuRhWT/results/F300010069_L01_25_Synlin20CycleKODAmp+Pur_BC128_1_.R2.call.bam'
2023-06-27 14:57:59,755 - analysis.variant_caller - INFO - Collecting read mismatch data. This may take some time...
2023-06-27 14:57:59,778 - analysis.variant_caller - INFO - Calling variants.
2023-06-27 14:57:59,778 - analysis.variant_caller - INFO - Writing results.
2023-06-27 14:58:00,054 - analysis.variant_caller - INFO - Completed variant calling workflow.
2023-06-27 14:58:00,054 - root - INFO - Completed call workflow.
2023-06-27 14:58:00,054 - root - INFO - Completed /home1/06285/ihoskins/miniconda3/envs/satmut_utils_test/bin/satmut_utils
```

What permissions does your output directory and or the copied reference FASTA have? 

Another thought: do you have an environment variable "$SCRATCH" defined?
```
echo "$SCRATCH"
```

**Comment by stevechen123** [comment]
Hi @ijhoskins.

After looking at your run command, I suspected that it might have been a PATH issue. Turns out that if I set absolute path for everything, including the FASTQ files, it works!

For reasons unknown, using relative paths results in the error.

Thanks so much for your help! I really appreciate it!

**Comment by ijhoskins** [comment]
Hi @stevechen123 

Glad you got it to work! Can you confirm you were using satmut_utils v1.0.3 by executing the following?
```
conda list | grep satmut
```

v1.0.2 had a bug where relative FASTQ paths caused an error during alignment. However, I have not previously had issues supplying relative paths for the reference files. If you were experiencing this with v1.0.3, sounds like a bug I will need to address.

**Comment by stevechen123** [comment]
Yes. I am using v1.0.3.

satmut-utils              1.0.3                    pypi_0    pypi

**Comment by ijhoskins** [comment]
Hi @stevechen123 thank you for uncovering a bug! I confirm I get the same error if I don't pass a full path to the output_dir argument. However, I was able to run with relative file paths for FASTQs and reference files.

**Comment by ijhoskins** [comment]
This bug is now fixed in v1.0.4.
